algopy_testing/src/algopy/primitives/biguint.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import annotations
+
+import functools
+from typing import TYPE_CHECKING
+
+from algopy.constants import MAX_UINT512
+from algopy.primitives.bytes import Bytes
+
+if TYPE_CHECKING:
+    from algopy.primitives.uint64 import UInt64
+
+# TypeError, ValueError are used for operations that are compile time errors
+# ArithmeticError and subclasses are used for operations that would fail during AVM execution
+
+
+@functools.total_ordering
+class BigUInt:
+    """
+    A python implementation of an AVM 64-bit unsigned integer
+    """
+
+    value: int  # underlying 'int' value representing the BigUInt
+
+    def __init__(self, value: UInt64 | int = 0) -> None:
+        self.value = _as_int512(value)
+
+    def __repr__(self) -> str:
+        return f"{self.value}u"
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+    # comparison operations should not error when comparing any uint type
+    # e.g. UInt64, BigUInt, arc4.UInt*, arc4.BigUint*
+    # however will raise a TypeError if compared to something that is not a numeric value
+    # as this would be a compile error when compiled to TEAL
+    def __eq__(self, other: object) -> bool:
+        return self.value == _as_int512(other)
+
+    def __lt__(self, other: BigUInt | UInt64 | int) -> bool:
+        return self.value < _as_int512(other)
+
+    def __bool__(self) -> bool:
+        return self.value != 0
+
+    def __add__(self, other: BigUInt | UInt64 | int) -> BigUInt:
+        return _checked_result(self.value + _as_int512(other), "+")
+
+    # the reflected dunder methods (__radd__, __rsub___, etc.) should only be called when the
+    # BigUInt is on the right hand side and an int on the left
```

Comment:
For BigUInt we can expected slightly different behaviour, when calling a reflected method, the left hand side could be an `int` or a `UInt64`, so we should also use `_as_int512` to convert other

Reply:
Actually I can see the implementation is doing that correctly, so just need to update this comment to avoid confusion

Reply:
I've updated the comment. I hope it is correct now. :)

Reply:
Comment looks good :)